docs/docs/aztec/writing_efficient_contracts.md

```diff
@@ -0,0 +1,358 @@
+---
+title: Writing Efficient Contracts
+sidebar_position: 2
+tags: [Developers, Contracts]
+---
+
+import Image from "@theme/IdealImage";
+
+## Writing functions
+
+On Ethereum L1, all data is public and all execution is completely reproducible. The Aztec L2 takes on the challenge of execution of private functions on private data. This is done client side, along with the generation of corresponding proofs, so that the network can verify the proofs and append any encrypted data/nullifiers (privacy preserving state update).
+
+This highlights a key difference with how public vs private functions are written.
+
+:::info Writing efficiently
+
+- **Public functions** can be written intuitively - optimising for execution/gas as one would for EVM L2s
+- **Private functions** are optimised differently, as they are compiled to a circuit to be proven locally (see [Thinking in Circuits](https://noir-lang.org/docs/explainers/explainer-writing-noir))
```

Comment:
do we want to also. add how there are DA costs too for private fns though wethat is out of scope rn